virttest/test_setup.py

```diff
@@ -97,6 +100,106 @@ class PolkitConfigCleanupError(PolkitConfigError):
     pass
 
 
+class Setuper(object):
+
+    """
+    Virtual base abstraction of setuper.
+    """
+
+    __metaclass__ = ABCMeta
+
+    #: Skip the cleanup when error occurs
+    skip_cleanup_on_error = False
+
+    def __init__(self, test, params, env):
+        """
+        Initialize the setuper.
+
+        :param test: VirtTest instance.
+        :param params: Dictionary with the test parameters.
+        :param env: Dictionary with test environment.
+        """
+        self.test = test
+        self.params = params
+        self.env = env
+
+    @abstractmethod
+    def setup(self):
+        """Setup procedure."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def cleanup(self):
+        """Cleanup procedure."""
+        raise NotImplementedError
+
+
+class SetupManager(object):
+
+    """
+    Setup Manager implementation.
+
+    The instance can help do the setup stuff before test started and
+    do the cleanup stuff after test finished. This setup-cleanup
+    combined stuff will be performed in LIFO order.
+    """
+
+    def __init__(self):
+        self.__setupers = []
+        self.__setup_args = None
+
+    def initialize(self, test, params, env):
+        """
+        Initialize the setup manager.
+
+        :param test: VirtTest instance.
+        :param params: Dictionary with the test parameters.
+        :param env: Dictionary with test environment.
+        """
+        self.__setup_args = (test, params, env)
```

Comment:
should we also have to register the class here,
```python
self.register(ABCMeta)
```

Reply:
No, it's meaningless to register the metaclass.

Reply:
Okay, but we have to register before performing `setup_manager.do_setup()` ?

Reply:
Yes, it requires that. The reason why it is not designed to register `setuper`s statically is because we have the scenarios to register them conditionally, and some of the check must be performed on runtime.
```python
if params.get("setup_ksm", "no") == "yes":
    manager.register(KSMSetuper)
```

Reply:
fine, got it. Thanks for explaning.